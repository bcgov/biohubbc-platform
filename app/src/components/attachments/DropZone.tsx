import { mdiTrayArrowUp } from '@mdi/js';
import Icon from '@mdi/react';
import { Theme } from '@mui/material';
import Box from '@mui/material/Box';
import Link from '@mui/material/Link';
import Typography from '@mui/material/Typography';
import { makeStyles } from '@mui/styles';
import { ConfigContext } from 'contexts/configContext';
import { useContext } from 'react';
import Dropzone, { FileRejection } from 'react-dropzone';

const useStyles = makeStyles((theme: Theme) => ({
  dropZoneIcon: {
    color: theme.palette.text.secondary
  }
}));

const BYTES_PER_MEGABYTE = 1048576;

export interface IDropZoneProps {
  /**
   * Function called when files are accepted/rejected (via either drag/drop or browsing).
   *
   * Note: Files may be rejected due of file size limits or file number limits
   *
   * @memberof IDropZoneProps
   */
  onFiles: (acceptedFiles: File[], rejectedFiles: FileRejection[]) => void;
}

export interface IDropZoneConfigProps {
  /**
   * Maximum file size allowed (in bytes).
   *
   * @type {number}
   * @memberof IDropZoneProps
   */
  maxFileSize?: number;
  /**
   * Maximum number of files allowed.
   *
   * @type {number}
   * @memberof IDropZoneProps
   */
  maxNumFiles?: number;
  /**
   * Allow selecting multiple files while browsing.
   * Default: true
   *
   * Note: Does not impact drag/drop.
   *
   * @type {boolean}
   * @memberof IDropZoneProps
   */
  multiple?: boolean;
  /**
   * Comma separated list of allowed file extensions.
   *
   * Example: `'.pdf, .txt'`
   *
   * @type {string}
   * @memberof IDropZoneConfigProps
   */
  acceptedFileExtensions?: string;
}

export const DropZone: React.FC<React.PropsWithChildren<IDropZoneProps & IDropZoneConfigProps>> = (props) => {
  const classes = useStyles();
  const config = useContext(ConfigContext);

  const maxNumFiles = props.maxNumFiles || config?.MAX_UPLOAD_NUM_FILES;
  const maxFileSize = props.maxFileSize || config?.MAX_UPLOAD_FILE_SIZE;
  const multiple = props.multiple ?? true;
  const acceptedFileExtensions = props.acceptedFileExtensions;

  return (
    <Box className="dropZoneContainer">
      <Dropzone
        maxFiles={maxNumFiles}
        maxSize={maxFileSize}
        multiple={multiple}
        onDrop={props.onFiles}
        accept={props.acceptedFileExtensions}>
        {({ getRootProps, getInputProps }) => (
          <Box {...getRootProps()}>
            <input {...getInputProps()} data-testid="drop-zone-input" />
            <Box display="flex" flexDirection="column" alignItems="center" pt={4} pb={5} px={2}>
              <Icon className={classes.dropZoneIcon} path={mdiTrayArrowUp} size={1.5} />
              <Box
                sx={{
                  mt: 1,
                  fontSize: '1.125rem',
                  fontWeight: 700
                }}
              >
                Drag your {(multiple && 'files') || 'file'} here, or <Link underline="always">Browse Files</Link>
              </Box>
              <Box textAlign="center"
                sx={{
                  '& span + span': {
                    ml: 2
                  }
                }}
              >
                {acceptedFileExtensions && (
                  <Typography component="span" variant="subtitle2" color="textSecondary">
                    {`Supported Files: ${acceptedFileExtensions}`}
                  </Typography>
                )}
                {!!maxFileSize && maxFileSize !== Infinity && (
                  <Typography component="span" variant="subtitle2" color="textSecondary">
                    {`Max File Size: ${Math.round(maxFileSize / BYTES_PER_MEGABYTE)} MB`}
                  </Typography>
                )}
                {!!maxNumFiles && (
                  <Typography component="span" variant="subtitle2" color="textSecondary">
                    {`Max Files: ${maxNumFiles}`}
                  </Typography>
                )}
              </Box>
            </Box>
          </Box>
        )}
      </Dropzone>
    </Box>
  );
};

export default DropZone;
