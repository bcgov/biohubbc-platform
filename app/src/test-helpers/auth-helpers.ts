import { SYSTEM_ROLE } from 'constants/roles';
import { IAuthState } from 'contexts/authStateContext';
import Keycloak from 'keycloak-js';

export const UnauthenticatedUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: false
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: false,
    systemRoles: [],
    hasSystemRole: () => false,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    }
  }
};

export const SystemUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [],
    hasSystemRole: () => false,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    }
  }
};

export const SystemAdminAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN],
    hasSystemRole: () => true,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    },
    getLoginUrl: () => 'test/login'
  }
};

// Same effect as `Partial` but applies to all levels of a nested object
type Subset<T> = {
  [P in keyof T]?: T[P] extends Record<any, any> | undefined ? Subset<T[P]> : T[P];
};

/**
 * Build and return a mock auth state object.
 *
 * @param {{ base: IAuthState; overrides?: Subset<IAuthState> }} options
 * @return {*}  {IAuthState}
 */
export const getMockAuthState = (options: { base: IAuthState; overrides?: Subset<IAuthState> }): IAuthState => {
  const { base, overrides } = options;

  return {
    ...base,
    ...overrides,
    keycloakWrapper: {
      ...base.keycloakWrapper,
      ...overrides?.keycloakWrapper,
      Keycloak: {
        ...base.keycloakWrapper?.keycloak,
        ...overrides?.keycloakWrapper?.keycloak
      }
    }
  } as unknown as IAuthState;
};
