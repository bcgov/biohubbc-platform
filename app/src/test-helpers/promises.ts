/**
 * 
 */
export class Deferred<R = void, Q = void> {
    resolve: (value: R) => void = () => {};
    reject: (reason?: Q) => void = () => {};
    promise: Promise<R>;

    constructor() {
        this.promise = new Promise<R>((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });

        return this;
    }

    recycle = () => {
        this.promise = new Promise<R>((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });

        return this;
    }
}
