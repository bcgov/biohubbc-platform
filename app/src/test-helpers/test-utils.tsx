import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import appTheme from 'themes/appTheme'

const AllProviders: React.FC<React.PropsWithChildren> = (props) => {
  const { children} = props
  return (
    <ThemeProvider theme={appTheme}>    
      {children}
    </ThemeProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
