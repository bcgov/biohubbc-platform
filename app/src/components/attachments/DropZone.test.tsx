import { fireEvent, render, waitFor } from 'test-helpers/test-utils';
import DropZone from './DropZone';

const onFiles = jest.fn();

const renderContainer = () => {
  return render(<DropZone onFiles={onFiles} acceptedFileExtensions=".txt" />);
};

describe('DropZone', () => {
  it('renders component properly', () => {
    const { getByText } = renderContainer();

    expect(getByText("Browse Files", {exact: false})).toBeVisible();
  });

  it('calls the `onFiles` callback when files are selected', async () => {
    const { getByTestId } = renderContainer();

    const testFile = new File(['test png content'], 'testpng.txt', { type: 'text/plain' });

    const dropZoneInput = getByTestId('drop-zone-input');

    fireEvent.change(dropZoneInput, { target: { files: [testFile] } });

    await waitFor(() => {
      expect(onFiles).toHaveBeenCalledWith([testFile], [], expect.any(Object));
    });
  });
});
