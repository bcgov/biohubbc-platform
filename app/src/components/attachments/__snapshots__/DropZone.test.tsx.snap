// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropZone matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropZoneContainer MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-0"
      tabindex="0"
    >
      <input
        accept=".txt"
        autocomplete="off"
        data-testid="drop-zone-input"
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
      <div
        class="MuiBox-root css-ro5zyg"
      >
        <svg
          class="makeStyles-dropZoneIcon-2"
          role="presentation"
          style="width: 2.25rem; height: 2.25rem;"
          viewBox="0 0 24 24"
        >
          <path
            d="M2 12H4V17H20V12H22V17C22 18.11 21.11 19 20 19H4C2.9 19 2 18.11 2 17V12M12 2L6.46 7.46L7.88 8.88L11 5.75V15H13V5.75L16.13 8.88L17.55 7.45L12 2Z"
            style="fill: currentColor;"
          />
        </svg>
        <div
          class="makeStyles-dropZoneTitle-1 MuiBox-root css-1fhgjcy"
        >
          Drag your files here, or 
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-myvcsz-MuiTypography-root-MuiLink-root"
          >
            Browse Files
          </a>
        </div>
        <div
          class="MuiBox-root css-xi606m"
        >
          <div
            class="MuiBox-root css-0"
          >
            <span
              class="MuiTypography-root MuiTypography-subtitle2 css-75nrt7-MuiTypography-root"
            >
              Accepted files: .txt
            </span>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <span
              class="MuiTypography-root MuiTypography-subtitle2 css-75nrt7-MuiTypography-root"
            >
              Maximum file size: 50 MB
            </span>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <span
              class="MuiTypography-root MuiTypography-subtitle2 css-75nrt7-MuiTypography-root"
            >
              Maximum files: 10
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
